File: src/BackOffice/server/src/services/albums/domain/AlbumAggregate/Author.ts
import { ValueObject } from '../../../shared/domain';

export type AuthorProps = {
  AuthorId: string;
  Name: string;
  UserName: string;
};

class Author extends ValueObject<AuthorProps> {
  private _authorId: string;
  private _name: string;
  private _userName: string;

  public get authorId(): string {
    return this._authorId;
  }

  public get Name(): string {
    return this._name;
  }

  public get UserName(): string {
    return this._userName;
  }

  private constructor(props: AuthorProps) {
    super(props);

    this._authorId = props.AuthorId;
    this._name = props.Name;
    this._userName = props.UserName;
  }

  public static Create(
    AuthorId: string,
    Name: string,
    UserName: string
  ): Author {
    return new Author({ AuthorId, Name, UserName });
  }
}

export default Author;

File: src/Services/Ideas/MusicStore.Ideas.Domain/Shared/ValueObjects/Author.cs
using MusicStore.Shared.ValueObjects;

namespace MusicStore.Ideas.Domain.Shared.ValueObjects
{
    public class Author : ValueObject
    {
        public string Id { get; set; }
        public string Name { get; private set; }
        public string UserName { get; private set; }

        private Author(string id, string name, string userName)
        {
            Id = id;
            Name = name;
            UserName = userName;
        }

        public static Author Create(string id, string name , string userName)
        {
            return new Author(id, name, userName);
        }

        protected override IEnumerable<object> GetAtomicValues()
        {
            yield return Id;
            yield return Name;
            yield return UserName;
        }
    }
}
